### PR TITLE
Add BBE for `Regex` match operations

### PR DESCRIPTION
**Pull request by poorna2152**
## Purpose
Part of https://github.com/ballerina-platform/ballerina-dev-website/issues/9318

### CODE REVIEW

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    // Regular expression to extract production time (HH:MM) from a log string
+    string:RegExp timePattern = re `([01][0-9]|2[0-3]):([0-5][0-9])`;
+
+    // Find groups of the matching string from a specified starting index.
+    regexp:Groups timeMatchGroups = <regexp:Groups>timePattern.matchGroupsAt("Production time: 14:35", 17);
+    string hour = (<regexp:Span>timeMatchGroups[1]).substring();
+    string minutes = (<regexp:Span>timeMatchGroups[2]).substring();
+    io:println(string `Production hour: ${hour}`);
+    io:println(string `Production minute: ${minutes}`);
```
Comment (SasinduDilshara) [inline examples/regexp-match-operations/regexp_match_operations.bal:39]:
```suggestion
    io:println(string `Production minutes: ${minutes}`);
```

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    foreach string productCode in productCodes {
+        // Check if the product code fully matches the expected format.
+        boolean isValidProductCode = productCodePattern.isFullMatch(productCode);
+        io:println(string `Product Code: ${productCode}, Valid: ${isValidProductCode}`);
+
+        // If valid, extract the product type and unique ID from the match groups.
+        regexp:Groups? matchGroups = productCodePattern.fullMatchGroups(productCode);
+        if matchGroups is regexp:Groups {
+            string productType = (<regexp:Span>matchGroups[1]).substring();
```
Comment (SasinduDilshara) [inline examples/regexp-match-operations/regexp_match_operations.bal:19]: Lets add a comment saying what is the 0th element

File: examples/regexp-match-operations/regexp_match_operations.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# RegExp operations
+
+The ``RegExp`` type include a set of langlib functions for matching patterns in strings, enabling operations to find, validate, group, and extract data based on regular expressions.
```
Comment (SasinduDilshara) [inline examples/regexp-match-operations/regexp_match_operations.md:3]:
```suggestion
The `RegExp` type include a set of langlib functions for matching patterns in strings, enabling operations to find, validate, group, and extract data based on regular expressions.
```

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import ballerina/io;
+import ballerina/lang.regexp;
+
+public function main() {
+    string[] productCodes = ["ELEC-1234", "FURN-5678", "CLOT-1111", "FOOD-3333"];
+
+    // Regular expression to match and validate product codes.
+    // Format: [PRODUCT_TYPE]-[UNIQUE_ID]
```
Comment (SasinduDilshara) [inline examples/regexp-match-operations/regexp_match_operations.bal:8]: This comment should be after L4?

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    regexp:Span? productCode = productCodePattern.matchAt("Product code: FURN-5678", 14);
+    if productCode is regexp:Span {
+        io:println(string `Matched product: ${productCode.substring()}`);
+    }
+
+    // Regular expression to extract production time (HH:MM) from a log string
+    string:RegExp timePattern = re `([01][0-9]|2[0-3]):([0-5][0-9])`;
+
+    // Find groups of the matching string from a specified starting index.
+    regexp:Groups? timeMatchGroups = timePattern.matchGroupsAt("Production time: 14:35", 17);
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.bal:40]: May make sense to have two value that match the pattern in the string before and after the index to start at instead. That would demonstrate starting index better?

File: examples/regexp-match-operations/regexp_match_operations.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# RegExp operations
+
+The `RegExp` type include a set of langlib functions for matching patterns in strings, enabling operations to find, validate, group, and extract data based on regular expressions.
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.md:3]:
```suggestion
The `RegExp` type support a set of langlib functions to match patterns in strings and enable operations such as finding, validating, grouping, and extracting data based on regular expressions.
```

Double check grouping?

Comment (MaryamZi) [reply examples/regexp-match-operations/regexp_match_operations.md:3]: Also lang library vs langlib doesn't seem to be consistent across BBEs?

Comment (poorna2152) [reply examples/regexp-match-operations/regexp_match_operations.md]: Grouping is supported by match `fullMatchGroups` `matchGroupsAt` operations right?

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+        if !isValidProductCode {
+            continue;
+        }
+
+        // For a valid product code, extract the product type and unique ID from the match groups.
+        regexp:Groups? matchGroups = productCodePattern.fullMatchGroups(productCode);
+        if matchGroups is regexp:Groups {
+            // The first member in the `matchGroups` tuple is the entire matched string.
+            // The subsequent members represent the captured groups.
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.bal:24]:
```suggestion
            // The subsequent members represent the captured groups
```

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    // Regular expression to extract the time in the format `HH:MM` from a log string.
+    string:RegExp timePattern = re `([01][0-9]|2[0-3]):([0-5][0-9])`;
+
+    // Find groups of the matching string from a specified starting index.
+    regexp:Groups? timeMatchGroups = timePattern.matchGroupsAt(
+            "Production time: 14:35, Production time: 16:15", 41);
+    if timeMatchGroups is regexp:Groups {
+        string hour = (<regexp:Span>timeMatchGroups[1]).substring();
+        string minutes = (<regexp:Span>timeMatchGroups[2]).substring();
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.bal:46]: Can we avoid the casts?

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        io:println("Production hour: ", hour);
+        io:println("Production minutes: ", minutes);
+    }
+}
+
+function extractStringFromMatchGroup(regexp:Span? span) returns string {
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.bal:52]: See the comment on the other PR about using ensureType instead.

File: examples/regexp-match-operations/regexp_match_operations.md [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# RegExp operations
+
+The `RegExp` type support a set of langlib functions to match patterns in strings and enable operations such as finding, validating, grouping, and extracting data based on regular expressions.
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.md:3]:
```suggestion
The `RegExp` type supports a set of langlib functions to match patterns in strings and enable operations such as finding, validating, grouping, and extracting data based on regular expressions.
```

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    // Regular expression to match and validate product codes.
+    // Format: [PRODUCT_TYPE]-[UNIQUE_ID]
+    string:RegExp productCodePattern = re `^([A-Z]{4})-(\d{4})$`;
+
+    foreach string productCode in productCodes {
+        // Check if the product code fully matches the expected format.
+        boolean isValidProductCode = productCodePattern.isFullMatch(productCode);
+        io:println(string `Product Code: ${productCode}, Valid: ${isValidProductCode}`);
+
+        if !isValidProductCode {
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.bal:16]: Shall we add a comment here?

File: examples/regexp-match-operations/regexp_match_operations.bal [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            continue;
+        }
+
+        // For a valid product code, extract the product type and unique ID from the match groups.
+        regexp:Groups? matchGroups = productCodePattern.fullMatchGroups(productCode);
+        if matchGroups is regexp:Groups {
+            // The first member in the `matchGroups` tuple is the entire matched string.
+            // The subsequent members represent the captured groups
+            // (assigned to `productType` and `uniqueID` respectively).
```
Comment (MaryamZi) [inline examples/regexp-match-operations/regexp_match_operations.bal:25]:
```suggestion
            // corresponding to product type and unique ID respectively.
```

**Review by gimantha — APPROVED** [review]

**Pull request merged**